### PR TITLE
[zify] bugfix

### LINKED ISSUE #11089: lia fails with aliased type but omega succeeds
#### Description of the problem
```
Require Import Omega Lia.
Definition t := nat.
Goal forall (n : t), n = n.
Proof.
  intros.
  Fail lia.
  omega.
Qed.
Goal let t' := nat in forall (n : t'), n = n.
Proof.
  intros.
  Fail lia.
  omega.
Qed.
```
`lia` suceeds in both cases if `t` or `t'` is unfolded, but `omega` seems to be able to figure out the types without that step.

#### Coq Version
8.8 - master. `lia` and `omega` both fail on versions before that.

### PR DESCRIPTION
**Pull request by fajb**
- Use Tactics.apply to `zify` conclusion (instead of rewrite).
- Improve error reporting of  `Add Zify X`

**Kind:**  bug fix 

Closes #11089 , #14043

- [x] updated test-suite
- [x] updated documention

### CODE REVIEW

**Comment by mrhaandi** [comment]
Just out of curiosity: why `apply` and not a `refine`/`exact` variant?

**Comment by fajb** [comment]
> Just out of curiosity: why `apply` and not a `refine`/`exact` variant?
 
For simplicity.  I do not master the interface of `refine`.
Anyway, `apply` seems more robust than `rewrite` and probably not slower.

**Comment by pi8027** [comment]
@fajb I confirm this PR fixes mczify. Thanks!

**Comment by fajb** [comment]
@vbgl could you look at it? Thanks.

File: doc/sphinx/addendum/micromega.rst [hunk 1]
```diff
@@ -295,7 +295,7 @@ obtain :math:`-1`. By Theorem :ref:`Psatz <psatz_thm>`, the goal is valid.
    The :tacn:`zify` tactic can be extended with new types and operators by declaring and registering new typeclass instances using the following commands.
    The typeclass declarations can be found in the module ``ZifyClasses`` and the default instances can be found in the module ``ZifyInst``.
 
-.. cmd:: Add Zify @add_zify @one_term
+.. cmd:: Add Zify @add_zify @qualid
```
Comment (Zimmi48) [inline doc/sphinx/addendum/micromega.rst:298]: If you really want to restrict what is accepted by `Add Zify`, why didn't you change the parsing rules for the command? If there's a good reason for restricting this not at parsing time but only at interpretation time, then we will need to make a special exception for this in `doc/tools/docgram/common.edit_mlg`.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]

File: doc/sphinx/addendum/micromega.rst (hunk 1 above)
Comment (fajb) [reply doc/sphinx/addendum/micromega.rst:298]: Actually, I did in another recent PR but forgot to update the doc. 
You can see that it is now a `reference` in the mlg code (see https://github.com/coq/coq/blob/master/plugins/micromega/g_zify.mlg)

Comment (Zimmi48) [reply doc/sphinx/addendum/micromega.rst:298]: OK, great. Thanks for clarifying!

File: doc/sphinx/addendum/micromega.rst [hunk 2]
```diff
@@ -304,6 +304,9 @@ obtain :math:`-1`. By Theorem :ref:`Psatz <psatz_thm>`, the goal is valid.
       | {| PropOp | PropBinOp | PropUOp | Saturate }
 
    Registers an instance of the specified typeclass.
+   The typeclass type (e.g. :g:`BinOp Z.mul` or :g:`BinRel (@eq Z)`) has the additional constraint that
+   the non-implicit argument (here, :g:`Z.mul` or :g:`(@eq Z)`)
+   is either a @reference (here, :g:`Z.mul`) or the application of a @reference (here, :g:`@eq`) to a sequence of @one_term.
```
Comment (Zimmi48) [inline doc/sphinx/addendum/micromega.rst:309]:
```suggestion
   is either a :n:`@reference` (here, :g:`Z.mul`) or the application of a :n:`@reference` (here, :g:`@eq`) to a sequence of :n:`@one_term`.
```

**Review by Zimmi48 — APPROVED** [review]

File: theories/micromega/ZifyClasses.v [hunk 3]
```diff
@@ -7,7 +7,6 @@
 (*         *     GNU Lesser General Public License Version 2.1          *)
 (*         *     (see LICENSE file for the text of the license)         *)
 (************************************************************************)
-Set Primitive Projections.
```
Comment (vbgl) [inline theories/micromega/ZifyClasses.v:10]: Is this change intended? Is it related to bug being fixed?

Comment (fajb) [reply theories/micromega/ZifyClasses.v:10]: This is intended and is a quick fix for `mczify`.
The consequence is that projections are now  (for free) GlobRef.t and can therefore be directly indexed by `zify`.

Comment (vbgl) [reply theories/micromega/ZifyClasses.v:10]: Thanks for the clarification.

**Review by vbgl — APPROVED** [review]

**Comment by fajb** [comment]
Thanks a lot @vbgl .
